src/lib-components/forms/BaseInput.vue

```diff
@@ -5,34 +5,46 @@ import InputError from "./InputError.vue"
 import {
   useInputField,
   defaultInputProps,
+  defaultModelOpts,
   emailPattern,
   looseToNumber,
   phonePattern,
 } from "@/composables/forms"
 import type { TextLikeInput } from "@/composables/forms"
-import { computed, ref } from "vue"
+import { computed, useTemplateRef } from "vue"
 
 defineOptions({
   inheritAttrs: false,
 })
 
 const props = withDefaults(defineProps<TextLikeInput>(), defaultInputProps)
+const modelState = defineModel<TextLikeInput["modelValue"]>({
+  ...defaultModelOpts,
+  // NOTE(spk): transform with caution.  The emitted value will be locally updated
+  // in this component and applied to the input's value attribute.
+  //
+  // The upside to this side effect for number inputs is that we can enforce allowing
+  // only character sets that can be parsed into numbers across browsers (Firefox).
+  //
+  // Example: calling v.trim() may prevent the use of the space character by the user.
+  //
+  // https://github.com/vuejs/core/commit/de174e1aa756508c7542605a448e55a373afb1ed#diff-f1e0971afb7e61e75d37d0fc7d7f993c9681dddb7c689cb455d217a246b3d6f0
+  set: (v) => {
+    if (props.type === "number") {
+      return looseToNumber(v)
+    }
 
-defineEmits(["update:modelValue", "update:error"])
```

Comment:
`defineModel` handles declaring `update:modelValue` for you, so you can skip that step.  `"update:error"` is a crusty and is not an event we've been supporting, so you'll see all of them being removed from the input components.